File: solutions/nim/grade-school/1/grade_school.nim
# Dedicated to Shree DR.MDD
import std/[algorithm, sequtils]

type
  Student* = object
    personName*: string
    classLevel*: int

  School* = object
    studentList*: seq[Student]

proc `<`(left, right: Student): bool =
  return (left.classLevel, left.personName) < (right.classLevel, right.personName)

proc roster*(school: School): seq[string] =
  return school.studentList.sorted.mapIt(it.personName)

proc addStudent*(school: var School, personName: string, classLevel: int) =
  if school.studentList.filterIt(it.personName == personName) != []:
    raise ValueError.new()
  school.studentList.add(Student(personName: personName, classLevel: classLevel))

proc grade*(school: School, classLevel: int): seq[string] =
  return school.studentList.filterIt(it.classLevel == classLevel).sorted.mapIt(it.personName)

File: solutions/go/grade-school/1/grade_school.go
// For my Shree DR.MDD
package school

import (
	"sort"
)

type Grade struct {
	level   int
	students []string
}

type School struct {
	grades []Grade
}

func New() *School {
	return &School{}
}

func (g *Grade) addStudent(name string) {
	pos := sort.SearchStrings(g.students, name)
	g.students = append(g.students, "")
	copy(g.students[pos+1:], g.students[pos:])
	g.students[pos] = name
}

func (s *School) Add(name string, level int) {
	pos := sort.Search(len(s.grades), func(i int) bool { return s.grades[i].level >= level })
	if len(s.grades) > pos && s.grades[pos].level == level {
		s.grades[pos].addStudent(name)
	} else {
		s.grades = append(s.grades, Grade{})
		copy(s.grades[pos+1:], s.grades[pos:])
		s.grades[pos] = Grade{level, []string{name}}
	}
}

func (s *School) Grade(level int) []string {
	pos := sort.Search(len(s.grades), func(i int) bool { return s.grades[i].level >= level })
	if len(s.grades)-pos > 0 && s.grades[pos].level == level {
		return s.grades[pos].students
	}
	return []string{}
}

func (s *School) Enrollment() []Grade {
	return s.grades
}
